packages/http-client-csharp/generator/Microsoft.Generator.CSharp/src/Providers/ModelProvider.cs

```diff
@@ -527,23 +527,32 @@ p.Property is null
                     {
                         return GetUnknownDiscriminatorExpression(discriminator);
                     }
-                    else
+
+                    if (!type.IsFrameworkType && type.IsEnum)
                     {
-                        if (!type.IsFrameworkType && type.IsEnum && _inputModel.BaseModel.DiscriminatorProperty!.Type is InputEnumType inputEnumType)
+                        if (_inputModel.BaseModel.DiscriminatorProperty!.Type is InputEnumType inputEnumType)
                         {
-                            /* TODO: when customize the discriminator type to a enum, then we may not be able to get the correct TypeProvider in this way.
-                             * We will handle this when issue https://github.com/microsoft/typespec/issues/4313 is resolved.
-                             * */
                             var discriminatorProvider = CodeModelPlugin.Instance.TypeFactory.CreateEnum(enumType: inputEnumType);
-                            var enumMember = discriminatorProvider!.EnumValues.FirstOrDefault(e => e.Value.ToString() == _inputModel.DiscriminatorValue) ?? throw new InvalidOperationException($"invalid discriminator value {_inputModel.DiscriminatorValue}");
+                            var enumMember = discriminatorProvider!.EnumValues.FirstOrDefault(e => e.Value.ToString() == _inputModel.DiscriminatorValue)
+                                ?? throw new InvalidOperationException($"invalid discriminator value {_inputModel.DiscriminatorValue}");
                             /* {KindType}.{enumMember} */
-                            return TypeReferenceExpression.FromType(type).Property(enumMember.Name);
+                            return Static(type).Property(enumMember.Name);
                         }
-                        else
+
+                        // Handle custom fixed enum discriminator
+                        if (discriminator.CustomProvider?.Value?.Type is { IsEnum: true, IsValueType: true, IsStruct: false })
```

Comment:
I don't think we need anything more than IsEnum here since this is using the custom code type which will only be considered an enum for fixed enums.